pkg/rulefmt/rulefmt.go

```diff
@@ -101,29 +106,44 @@ type Rule struct {
 	Annotations map[string]string `yaml:"annotations,omitempty"`
 }
 
+// RuleNode adds yaml.v3 layer to support line and column outputs for invalid rules.
+type RuleNode struct {
+	Record      yaml.Node         `yaml:"record,omitempty"`
+	Alert       yaml.Node         `yaml:"alert,omitempty"`
+	Expr        yaml.Node         `yaml:"expr"`
+	For         model.Duration    `yaml:"for,omitempty"`
+	Labels      map[string]string `yaml:"labels,omitempty"`
+	Annotations map[string]string `yaml:"annotations,omitempty"`
+}
+
 // Validate the rule and return a list of encountered errors.
-func (r *Rule) Validate() (errs []error) {
-	if r.Record != "" && r.Alert != "" {
-		errs = append(errs, errors.Errorf("only one of 'record' and 'alert' must be set"))
+func (r *RuleNode) Validate() (errs []error) {
+	if r.Record.Value != "" && r.Alert.Value != "" {
+		errs = append(errs, errors.Errorf("only one of 'record' and 'alert' must be set, lines: %d %d, columns: %d %d", r.Record.Line, r.Alert.Line, r.Record.Column, r.Alert.Column))
 	}
-	if r.Record == "" && r.Alert == "" {
-		errs = append(errs, errors.Errorf("one of 'record' or 'alert' must be set"))
+	if r.Record.Value == "" && r.Alert.Value == "" {
+		if r.Record.Value == "0" {
+			errs = append(errs, errors.Errorf("one of 'record' or 'alert' must be set, lines: %d, columns: %d", r.Alert.Line, r.Alert.Column))
+		} else {
+			errs = append(errs, errors.Errorf("one of 'record' or 'alert' must be set, lines: %d, columns: %d", r.Record.Line, r.Record.Column))
+		}
 	}
 
-	if r.Expr == "" {
-		errs = append(errs, errors.Errorf("field 'expr' must be set in rule"))
-	} else if _, err := promql.ParseExpr(r.Expr); err != nil {
-		errs = append(errs, errors.Wrap(err, "could not parse expression"))
+	if r.Expr.Value == "" {
+		fmt.Println(r.Expr.Line)
+		errs = append(errs, errors.Errorf("field 'expr' must be set in rule, line: %d, column: %d", r.Expr.Line, r.Expr.Column))
+	} else if _, err := promql.ParseExpr(r.Expr.Value); err != nil {
+		errs = append(errs, errors.Wrapf(err, "could not parse expression, line: %d, column: %d", r.Expr.Line, r.Expr.Column))
```

Comment:
This should end up looking quite strange, since the error returned by the parser will include position information, too.

Normally the parser returns a `promql.ParseErr`. This struct already includes the position information.

What the PromQL language server does here, is to add the line and column where the expression starts to the line and column in the error message and then simply convert the error message to a string

Reply:
We need the position coordinates to make debugging rules easier for the user, in case of a larger file, increasing the convenience. Similar behaviour is noted in various format checkers as well.

Reply:
That's clear.

The problem with the current approach is that the output will look roughly like
`could not parse expression, line: 3, column: 10: 2:14: Unexpected identifier in label matcher, expected string`. 

It would be better if the output looked like:
`rules.yaml:4:24: could not parse expression: Unexpected identifier in label matcher, expected string`. Note that the positions have been added together here.

Reply:
The current implement is, (from live example)

```
rules.yml: Groupname should not be empty, line: 2, column: 5
rules.yml: group "recording-node-exporter", rule 1, "heap": field 'expr' must be set in rule, line: 13, column: 12
```
is this acceptable? @brian-brazil

Reply:
can you post your rules file here? @slrtbtfs

Reply:
This wasn't an actual output, but rather behaviour guessed from the code. The cases where the problem I described should occur, is if there is a non-empty, syntactically wrong query inside the file being checked.

Reply:
I've constructed a real example:

Using this `rules.yaml`:

```
groups:
    - name: recording
      rules:
          - expr: |
              go_goroutines}
            record: foo
```

The output with your patches applied is:

```
Checking rules.yaml
  FAILED:
rules.yaml: group "recording", rule 0, "foo": could not parse expression, line: 4, column: 19: 1:14: parse error: could not parse remaining input "}\n"...
```

Reply:
thanks. I will have a look over this.

Reply:
I have updated. The current version gives the following output for the above config,

```
Checking ../../rules.yml
  FAILED:
../../rules.yml: 6:21 group "recording", rule 0, "foo": 4:19 could not parse expression: 1:14: parse error: could not parse remaining input "}\n"...

```

and for the initial example ,
```
  FAILED:
../../rules.yml: 2:5 Groupname should not be empty
../../rules.yml: 12:15 group "recording-node-exporter", rule 1, "heap": 13:12 field 'expr' must be set in rule
```

Reply:
Thanks. Looks better now.

---

> `../../rules.yml: 2:5`

That space should not be there and there should be a colon at the end of the position. To be consistent with conventions from various compilers the position should be of the form
`../../rules.yml:2:5:`

(nit)

----

> `../../rules.yml: 5:21 group "recording", rule 0, "foo": 4:19 could not parse expression: 1:14: parse error: could not parse remaining input "}\n"...`

This one looks really confusing, since a single error message contains three different position coordinates and a verbal description of the position.

Ideally there would be only one position, directly at the beginning of the error line. Something like

 `../../rules.yml:5:28: group "recording", rule 0, "foo": could not parse expression: parse error: could not parse remaining input "}\n"...`

Maybe even something more customized like:
`../../rules.yml:5:28: Failed to parse PromQL expression: parse error: could not parse remaining input "}\n"...`

Reply:
As the design of the rulefmt, this looks tough. If we remove the parser checks and let it handle by itself as earlier, then this can be possible, but again on the right-hand side, towards the ending of the sentence.

Reply:
It should be possible by just typecasting the `err` result of `promql.ParseExpr()` to a `promql.ParseErr` and the using the fields in that data structure.

Reply:
The space before the number is due to `errors.Wrapf()`. I don't think there is any way to remove this space. Any idea @slrtbtfs ?

Leaving that, rest I have given a new implementation of the logic. That solves most of what you said.

Reply:
By not using `Wrapf()` but instead constructing a custom error message from the data contained in the error.